libs/azure/odc/azure/__init__.py

```diff
@@ -0,0 +1,56 @@
+"""Azure blob storage crawling and YAML fetching utilities
+"""
+from azure.storage.blob import ContainerClient, BlobClient
+from typing import List, Tuple, Optional
+from multiprocessing.dummy import Pool as ThreadPool
+from functools import partial
+
+def find_blobs(account_url: str, container_name: str, credential: str, prefix: str, suffix: str):
+    blob_list = []
+    container = ContainerClient(account_url=account_url, container_name=container_name, credential=credential)
+    for blob_record in container.list_blobs(name_starts_with=prefix):
+        blob_name = blob_record['name']
+
+        if blob_name.endswith(suffix):
+            blob_list.append(blob_name)
+    return blob_list
+
+
+def download_yamls(account_url: str, container_name: str, credential: str, yaml_urls: List[str],
+                   workers: int = 32) -> List[Tuple[Optional[bytes], str, Optional[str]]]:
+    """Download all YAML's in a list of blob names and generate content
+    Arguments:
+        account_url {str} -- Azure account url
+        container_name {str} -- Azure container name
+        credential {str} -- Azure credential token
+
+        yaml_urls {list} -- List of URL's to download YAML's from
+        workers {int} -- Number of workers to use for Thredds Downloading
+    Returns:
+        list -- tuples of contents and filenames
+    """
+    # use a threadpool to download from Azure blobstore
+
+    pool = ThreadPool(workers)
+    yamls = pool.map(partial(_download, account_url, container_name, credential), yaml_urls)
+    pool.close()
+    pool.join()
+    return yamls
+
+
+def _download(account_url: str, container_name: str, credential: str,
+              blob_name: str) -> Tuple[Optional[bytes], str, Optional[str]]:
+    """Internal method to download YAML's from Azure via BlobClient
+    Arguments:
+        account_url {str} -- Azure account url
+        container_name {str} -- Azure container name
+        credential {str} -- Azure credential token
+        blob_name {str} -- Blob name to download
+    Returns:
+        tuple -- URL content, target file and placeholder for error
+    """
+
+    blob = BlobClient(account_url=account_url, container_name=container_name, credential=credential,
```

Comment:
Should this be guarded with `try: except`? Function signature suggests error as return value.

Reply:
by "this" I mean including `return ..` line

Reply:
This was some bad copy-pasta from the THREDDs lib. That error placeholder was for capturing HTTP errors. I have removed this. With the updated the function signature, is this still required?

Reply:
This signature still suggests "error as value" result with `(None, url)` indicating failure for a given url. And I think that's what you want, given that you are using `pool.map` and probably don't want the entire lot to fail when one blob fails to fetch. 

If this just raises exception on error then signature should be just `url:str -> bytes` or exception if anything goes wrong.

I would also put `blob_name` as the first argument, you can call `pool.map` with lambda, in case this order is motivated by the use of `partial` in code that calls this, or use named arguments with `partial`

```python
pool.map(partial(_download, 
        account_url=account_url, 
        conatiner_name=container_name, 
        credential=credential), yaml_urls)
```

Reply:
Also is there a standard way to represent `container_name+blob_name` as a url? Is `account_url` + `credentials` related, or is it more like `account_url + container_name + blob_name`?